benchmark-utils_test.go

```diff
@@ -224,21 +201,62 @@ func runGetObjectBenchmark(b *testing.B, obj ObjectLayer, objSize int) {
 
 }
 
-// randomly picks a character and returns its equivalent byte array.
-func getRandomByte() []byte {
-	const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
-	// seeding the random number generator.
-	rand.Seed(time.Now().UnixNano())
-	var b byte
-	// pick a character randomly.
-	b = letterBytes[rand.Intn(len(letterBytes))]
-	return []byte{b}
+const alphabets = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+// Returns random character as byte from letterBytes.
+func getRandByte() byte {
+	return alphabets[rand.Intn(len(alphabets))]
 }
 
-// picks a random byte and repeats it to size bytes.
-func generateBytesData(size int) []byte {
-	// repeat the random character chosen size
-	return bytes.Repeat(getRandomByte(), size)
+// Returns os.File and MD5 hash string of required size of random bytes.
+func mustGetRandomReader(size int) (file *os.File, hash string) {
+	rand.Seed(time.Now().UnixNano())
+	hasher := md5.New()
+
+	// Create temporary file.
+	tmpfile, err := ioutil.TempFile("", "minio-bench")
+	if err != nil {
+		panic(err)
+	}
+	defer os.Remove(tmpfile.Name())
+
+	var buf []byte
+	if size < readSizeV1 {
+		buf = bytes.Repeat([]byte{getRandByte()}, size)
+	} else {
+		buf = bytes.Repeat([]byte{getRandByte()}, readSizeV1)
+	}
+
+	// Write to the temporary file.
+	bytesLeft := size
+	n := 0
+	for bytesLeft > 0 {
+		if bytesLeft > readSizeV1 {
+			n = readSizeV1
+		} else {
+			n = bytesLeft
+		}
+
+		if _, err = tmpfile.Write(buf[:n]); err != nil {
```

Comment:
I made below changes to the function `mustGetRandomReader()`
```
func mustGetRandomReader(size int) (file *os.File, hash string) {
	rand.Seed(time.Now().UnixNano())
	hasher := md5.New()

	// Create temporary file.
	tmpfile, err := ioutil.TempFile("", "minio-bench")
	if err != nil {
		panic(err)
	}
	defer os.Remove(tmpfile.Name())

	var buf []byte
	if size < readSizeV1 {
		buf = bytes.Repeat([]byte{getRandByte()}, size)
	} else {
		buf = bytes.Repeat([]byte{getRandByte()}, readSizeV1)
	}

+	writer := bufio.NewWriter(tmpfile)

	// Write to the temporary file.
	bytesLeft := size
	n := 0
	for bytesLeft > 0 {
		if bytesLeft > readSizeV1 {
			n = readSizeV1
		} else {
			n = bytesLeft
		}

*		if _, err = writer.Write(buf[:n]); err != nil {
			panic(err)
		}

		hasher.Write(buf)

		bytesLeft -= n
	}

+	writer.Flush()

	if err = tmpfile.Close(); err != nil {
		panic(err)
	}

	// Open the temporary file for reading.
	f, err := os.Open(tmpfile.Name())
	if err != nil {
		panic(err)
	}

	return f, hex.EncodeToString(hasher.Sum(nil))
}
```

Here is the benchmark result.

```
$ go test -v -run=NONE -bench BenchmarkPutObject25MbFS | tee bench1
PASS
BenchmarkPutObject25MbFS-8             5         260697238 ns/op          379444 B/op       3446 allocs/op
ok      github.com/minio/minio  3.252s


$ go test -v -run=NONE -bench BenchmarkPutObject25MbFS | tee bench2
PASS
BenchmarkPutObject25MbFS-8             5         260629370 ns/op          379444 B/op       3446 allocs/op
ok      github.com/minio/minio  3.253s


$ benchcmp bench1 bench2  
benchmark                      old ns/op     new ns/op     delta
BenchmarkPutObject25MbFS-8     260697238     260629370     -0.03%

benchmark                      old allocs     new allocs     delta
BenchmarkPutObject25MbFS-8     3446           3446           +0.00%

benchmark                      old bytes     new bytes     delta
BenchmarkPutObject25MbFS-8     379444        379444        +0.00%

```

I also tried `bufio.NewReader()` by modifying function `runPutObjectBenchmark()`
```
// Benchmark utility functions for ObjectLayer.PutObject().
// Creates Object layer setup ( MakeBucket ) and then runs the PutObject benchmark.
func runPutObjectBenchmark(b *testing.B, obj ObjectLayer, objSize int) {
	var err error
	// obtains random bucket name.
	bucket := getRandomBucketName()
	// create bucket.
	err = obj.MakeBucket(bucket)
	if err != nil {
		b.Fatal(err)
	}

	// Get file and md5sum.
	file, md5Sum := mustGetRandomReader(objSize)
	defer file.Close()

	metadata := make(map[string]string)
	metadata["md5Sum"] = md5Sum

	reader := bufio.NewReader(file)

	// benchmark utility which helps obtain number of allocations and bytes allocated per ops.
	b.ReportAllocs()
	// the actual benchmark for PutObject starts here. Reset the benchmark timer.
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		// Move the file pointer to beginning of the file
		file.Seek(0, 0)
		reader.Reset(file)

		// insert the object.
		md5Sum, err = obj.PutObject(bucket, "object"+strconv.Itoa(i), int64(objSize), reader, metadata)
		if err != nil {
			b.Fatal(err)
		}
		if md5Sum != metadata["md5Sum"] {
			b.Fatalf("Write no: %d: Md5Sum mismatch during object write into the bucket: Expected %s, got %s", i+1, md5Sum, metadata["md5Sum"])
		}
	}
	// Benchmark ends here. Stop timer.
	b.StopTimer()
}
```

Here is the benchmark result.

```
$ go test -v -run=NONE -bench BenchmarkPutObject25MbFS | tee bench3
PASS
BenchmarkPutObject25MbFS-8             5         260963565 ns/op          379460 B/op       3446 allocs/op
ok      github.com/minio/minio  3.263s

$ benchcmp bench1 bench3
benchmark                      old ns/op     new ns/op     delta
BenchmarkPutObject25MbFS-8     260697238     260963565     +0.10%

benchmark                      old allocs     new allocs     delta
BenchmarkPutObject25MbFS-8     3446           3446           +0.00%

benchmark                      old bytes     new bytes     delta
BenchmarkPutObject25MbFS-8     379444        379460        +0.00%
```